<h1><center>PA2 Report</center></h1>

Group information:   
Course: CSE 574    
Group number: 29       
Xiyun Xie (50162104)     
Mollie Wugetemole (50165532)    
Yutong Yang (50321338)    
### Part 1:
#### Approach 1:
#### Output:

Accuracy of the model = 0.762

### Report 1:

The method we do classification is that we collect words in training reviews and put all of them into two piles(positive count and negative count) of words according to the given label of training data. Then we compute the ratio of occuracencies of each word in positive and negative classes and we take its log value, which means if the log value is much greater than 0, then a certain word is more likely to be considered as a positive word since it appears more in positive reviews. For each of our testing review, we count if words appears more as a positive word or not by computing the it's log ratio we previously computed with a kind of threshhold. We first set the threshold as (-0 , +0),  which means if the log ratio is greater than +0, we classify it as a positive word, negative otherwise. But we found the accuracy is very low(65%) and we think the reason is that there are too many vavague or neural words that just happens to appear more in positive and negative reviews and their log ratios are nearby 0. Since there are many of them, they will bring a lot of uncertainties to our classification model and cause mis-classifications. When we set threshold (-0.3, +0.3), the accuracy gets much better (76.2%, shown as our output above). And we are curious how the accuracy will change when we keep increasing the range of ratio of words we don't want to use. We find that the accuracy will be maximized with threshold (-1 , +1) and accuracy is 80.6%. It starts to drop when we keep increasing the range. We know everything depends on our training review, so the accuracy will vary a lot by different kind of reviews, but we at least find some patterns of the accuracy of our model.   

#### Approach 2:
#### Output:

(-0 , +0) 1 hidden layer and 10 units: 

In [ ]:
Train acc: 0.497833, Test_acc: 0.501250
Train acc: 0.499792, Test_acc: 0.500000
Train acc: 0.500125, Test_acc: 0.500000
Train acc: 0.500375, Test_acc: 0.500000
Train acc: 0.500375, Test_acc: 0.500000
Train acc: 0.500458, Test_acc: 0.500000
Train acc: 0.500458, Test_acc: 0.500000
Train acc: 0.500500, Test_acc: 0.500000
Train acc: 0.500500, Test_acc: 0.500000
Train acc: 0.500417, Test_acc: 0.500000
Train acc: 0.500583, Test_acc: 0.500000
Train acc: 0.500708, Test_acc: 0.500000
Train acc: 0.500792, Test_acc: 0.500000
Train acc: 0.500917, Test_acc: 0.501250
Train acc: 0.500917, Test_acc: 0.501250
Train acc: 0.500875, Test_acc: 0.501250
Train acc: 0.500792, Test_acc: 0.501250
Train acc: 0.501042, Test_acc: 0.502500
Train acc: 0.501500, Test_acc: 0.502500
Train acc: 0.501375, Test_acc: 0.502500
Train acc: 0.501583, Test_acc: 0.502500
Train acc: 0.501667, Test_acc: 0.502500
Train acc: 0.502042, Test_acc: 0.503750
Train acc: 0.503042, Test_acc: 0.505000
Train acc: 0.503833, Test_acc: 0.508750
Train acc: 0.544208, Test_acc: 0.760000
Train acc: 0.886333, Test_acc: 0.825000
Train acc: 0.897375, Test_acc: 0.830000
Train acc: 0.904958, Test_acc: 0.835000
Train acc: 0.910417, Test_acc: 0.838750
Train acc: 0.914250, Test_acc: 0.841250
Train acc: 0.916500, Test_acc: 0.837500
Train acc: 0.918958, Test_acc: 0.835000
Train acc: 0.919333, Test_acc: 0.838750
Train acc: 0.919417, Test_acc: 0.837500
Train acc: 0.917625, Test_acc: 0.842500
Train acc: 0.923167, Test_acc: 0.846250
Train acc: 0.919792, Test_acc: 0.846250
Train acc: 0.926000, Test_acc: 0.851250
Train acc: 0.922875, Test_acc: 0.845000
Train acc: 0.921958, Test_acc: 0.836250
Train acc: 0.924583, Test_acc: 0.852500
Train acc: 0.929125, Test_acc: 0.850000
Train acc: 0.930333, Test_acc: 0.845000
Train acc: 0.931000, Test_acc: 0.841250
Train acc: 0.929708, Test_acc: 0.845000
Train acc: 0.931417, Test_acc: 0.846250
Train acc: 0.933583, Test_acc: 0.845000
Train acc: 0.931125, Test_acc: 0.853750
Train acc: 0.934250, Test_acc: 0.855000
Time elapsed - 25.074923753738403 seconds.

(-0 , +0) 1 hidden layer and 50 units:

In [ ]:
Train acc: 0.496083, Test_acc: 0.500000
Train acc: 0.499958, Test_acc: 0.500000
Train acc: 0.499958, Test_acc: 0.500000
Train acc: 0.499958, Test_acc: 0.500000
Train acc: 0.499958, Test_acc: 0.500000
Train acc: 0.500000, Test_acc: 0.500000
Train acc: 0.499917, Test_acc: 0.500000
Train acc: 0.499917, Test_acc: 0.500000
Train acc: 0.499958, Test_acc: 0.500000
Train acc: 0.499958, Test_acc: 0.500000
Train acc: 0.499958, Test_acc: 0.500000
Train acc: 0.499958, Test_acc: 0.500000
Train acc: 0.499958, Test_acc: 0.500000
Train acc: 0.499958, Test_acc: 0.500000
Train acc: 0.499958, Test_acc: 0.500000
Train acc: 0.500000, Test_acc: 0.500000
Train acc: 0.499958, Test_acc: 0.500000
Train acc: 0.499958, Test_acc: 0.500000
Train acc: 0.499958, Test_acc: 0.500000
Train acc: 0.499958, Test_acc: 0.500000
Train acc: 0.500000, Test_acc: 0.500000
Train acc: 0.499958, Test_acc: 0.500000
Train acc: 0.500000, Test_acc: 0.500000
Train acc: 0.500000, Test_acc: 0.500000
Train acc: 0.500000, Test_acc: 0.500000
Train acc: 0.499958, Test_acc: 0.500000
Train acc: 0.499958, Test_acc: 0.500000
Train acc: 0.500042, Test_acc: 0.500000
Train acc: 0.500042, Test_acc: 0.500000
Train acc: 0.500042, Test_acc: 0.500000
Train acc: 0.500083, Test_acc: 0.500000
Train acc: 0.500167, Test_acc: 0.500000
Train acc: 0.500083, Test_acc: 0.500000
Train acc: 0.500292, Test_acc: 0.500000
Train acc: 0.500875, Test_acc: 0.502500
Train acc: 0.570708, Test_acc: 0.845000
Train acc: 0.864708, Test_acc: 0.840000
Train acc: 0.887875, Test_acc: 0.832500
Train acc: 0.888125, Test_acc: 0.847500
Train acc: 0.877958, Test_acc: 0.838750
Train acc: 0.888250, Test_acc: 0.827500
Train acc: 0.891542, Test_acc: 0.847500
Train acc: 0.892375, Test_acc: 0.843750
Train acc: 0.888917, Test_acc: 0.842500
Train acc: 0.895042, Test_acc: 0.841250
Train acc: 0.892458, Test_acc: 0.843750
Train acc: 0.894583, Test_acc: 0.840000
Train acc: 0.896542, Test_acc: 0.841250
Train acc: 0.900125, Test_acc: 0.837500
Train acc: 0.903417, Test_acc: 0.826250
Time elapsed - 27.42438530921936 seconds.

(-0 , +0) 2 hidden layer and 10 units each layer:

In [ ]:
Train acc: 0.586833, Test_acc: 0.596250
Train acc: 0.724167, Test_acc: 0.757500
Train acc: 0.811000, Test_acc: 0.790000
Train acc: 0.846875, Test_acc: 0.821250
Train acc: 0.866042, Test_acc: 0.832500
Train acc: 0.875375, Test_acc: 0.836250
Train acc: 0.888292, Test_acc: 0.845000
Train acc: 0.894083, Test_acc: 0.838750
Train acc: 0.902292, Test_acc: 0.836250
Train acc: 0.905125, Test_acc: 0.838750
Train acc: 0.904375, Test_acc: 0.838750
Train acc: 0.907625, Test_acc: 0.841250
Train acc: 0.913667, Test_acc: 0.840000
Train acc: 0.914000, Test_acc: 0.838750
Train acc: 0.914458, Test_acc: 0.846250
Train acc: 0.915667, Test_acc: 0.845000
Train acc: 0.916417, Test_acc: 0.842500
Train acc: 0.916917, Test_acc: 0.832500
Train acc: 0.915917, Test_acc: 0.848750
Train acc: 0.927417, Test_acc: 0.861250
Train acc: 0.929333, Test_acc: 0.851250
Train acc: 0.931083, Test_acc: 0.861250
Train acc: 0.935417, Test_acc: 0.847500
Train acc: 0.930875, Test_acc: 0.856250
Train acc: 0.933583, Test_acc: 0.862500
Train acc: 0.937958, Test_acc: 0.860000
Train acc: 0.939625, Test_acc: 0.860000
Train acc: 0.940292, Test_acc: 0.853750
Train acc: 0.940917, Test_acc: 0.857500
Train acc: 0.940917, Test_acc: 0.855000
Train acc: 0.942292, Test_acc: 0.852500
Train acc: 0.941458, Test_acc: 0.851250
Train acc: 0.944250, Test_acc: 0.853750
Train acc: 0.943167, Test_acc: 0.855000
Train acc: 0.938458, Test_acc: 0.850000
Train acc: 0.939042, Test_acc: 0.851250
Train acc: 0.938125, Test_acc: 0.852500
Train acc: 0.933667, Test_acc: 0.855000
Train acc: 0.928792, Test_acc: 0.857500
Train acc: 0.934458, Test_acc: 0.853750
Train acc: 0.940375, Test_acc: 0.861250
Train acc: 0.943125, Test_acc: 0.860000
Train acc: 0.943500, Test_acc: 0.853750
Train acc: 0.944042, Test_acc: 0.857500
Train acc: 0.945792, Test_acc: 0.848750
Train acc: 0.945458, Test_acc: 0.848750
Train acc: 0.948750, Test_acc: 0.856250
Train acc: 0.947917, Test_acc: 0.843750
Train acc: 0.949083, Test_acc: 0.847500
Train acc: 0.950250, Test_acc: 0.846250
Time elapsed - 24.950228691101074 seconds.

(-0 , +0) 2 hidden layer and 50 units each layer:

In [ ]:
Train acc: 0.564417, Test_acc: 0.626250
Train acc: 0.707833, Test_acc: 0.717500
Train acc: 0.778958, Test_acc: 0.785000
Train acc: 0.822083, Test_acc: 0.816250
Train acc: 0.842917, Test_acc: 0.815000
Train acc: 0.853083, Test_acc: 0.838750
Train acc: 0.869792, Test_acc: 0.831250
Train acc: 0.879292, Test_acc: 0.835000
Train acc: 0.889167, Test_acc: 0.846250
Train acc: 0.892375, Test_acc: 0.848750
Train acc: 0.896042, Test_acc: 0.847500
Train acc: 0.903458, Test_acc: 0.853750
Train acc: 0.907792, Test_acc: 0.832500
Train acc: 0.903792, Test_acc: 0.838750
Train acc: 0.909625, Test_acc: 0.843750
Train acc: 0.912000, Test_acc: 0.855000
Train acc: 0.911917, Test_acc: 0.857500
Train acc: 0.916125, Test_acc: 0.858750
Train acc: 0.919375, Test_acc: 0.855000
Train acc: 0.917708, Test_acc: 0.853750
Train acc: 0.919667, Test_acc: 0.855000
Train acc: 0.922167, Test_acc: 0.853750
Train acc: 0.924833, Test_acc: 0.861250
Train acc: 0.927250, Test_acc: 0.860000
Train acc: 0.929333, Test_acc: 0.865000
Train acc: 0.929208, Test_acc: 0.853750
Train acc: 0.931042, Test_acc: 0.850000
Train acc: 0.930167, Test_acc: 0.851250
Train acc: 0.931542, Test_acc: 0.853750
Train acc: 0.931042, Test_acc: 0.848750
Train acc: 0.934750, Test_acc: 0.858750
Train acc: 0.934292, Test_acc: 0.865000
Train acc: 0.937583, Test_acc: 0.845000
Train acc: 0.932250, Test_acc: 0.848750
Train acc: 0.935292, Test_acc: 0.852500
Train acc: 0.935542, Test_acc: 0.855000
Train acc: 0.937042, Test_acc: 0.850000
Train acc: 0.937208, Test_acc: 0.858750
Train acc: 0.937333, Test_acc: 0.848750
Train acc: 0.937958, Test_acc: 0.846250
Train acc: 0.937417, Test_acc: 0.842500
Train acc: 0.938792, Test_acc: 0.846250
Train acc: 0.939750, Test_acc: 0.846250
Train acc: 0.942125, Test_acc: 0.846250
Train acc: 0.940042, Test_acc: 0.856250
Train acc: 0.940708, Test_acc: 0.853750
Train acc: 0.941042, Test_acc: 0.852500
Train acc: 0.941167, Test_acc: 0.855000
Train acc: 0.941542, Test_acc: 0.853750
Train acc: 0.943000, Test_acc: 0.860000
Time elapsed - 28.478267431259155 seconds.

=============================================================================================

(-0.4 , +0.4) 1 hidden layer and 10 units: 

In [ ]:
Train acc: 0.684125, Test_acc: 0.771250
Train acc: 0.821792, Test_acc: 0.810000
Train acc: 0.856417, Test_acc: 0.816250
Train acc: 0.872125, Test_acc: 0.820000
Train acc: 0.882000, Test_acc: 0.825000
Train acc: 0.888875, Test_acc: 0.818750
Train acc: 0.894750, Test_acc: 0.827500
Train acc: 0.898833, Test_acc: 0.826250
Train acc: 0.903875, Test_acc: 0.832500
Train acc: 0.907417, Test_acc: 0.828750
Train acc: 0.910083, Test_acc: 0.830000
Train acc: 0.912583, Test_acc: 0.833750
Train acc: 0.914667, Test_acc: 0.833750
Train acc: 0.913625, Test_acc: 0.831250
Train acc: 0.915208, Test_acc: 0.838750
Train acc: 0.917792, Test_acc: 0.842500
Train acc: 0.916667, Test_acc: 0.840000
Train acc: 0.915333, Test_acc: 0.831250
Train acc: 0.913667, Test_acc: 0.825000
Train acc: 0.913750, Test_acc: 0.835000
Train acc: 0.916958, Test_acc: 0.832500
Train acc: 0.921833, Test_acc: 0.845000
Train acc: 0.924542, Test_acc: 0.842500
Train acc: 0.924292, Test_acc: 0.842500
Train acc: 0.923958, Test_acc: 0.846250
Train acc: 0.925250, Test_acc: 0.837500
Train acc: 0.925958, Test_acc: 0.836250
Train acc: 0.925625, Test_acc: 0.835000
Train acc: 0.926000, Test_acc: 0.832500
Train acc: 0.925333, Test_acc: 0.837500
Train acc: 0.923250, Test_acc: 0.838750
Train acc: 0.924500, Test_acc: 0.837500
Train acc: 0.924417, Test_acc: 0.836250
Train acc: 0.925958, Test_acc: 0.843750
Train acc: 0.926833, Test_acc: 0.848750
Train acc: 0.927375, Test_acc: 0.846250
Train acc: 0.927667, Test_acc: 0.847500
Train acc: 0.928333, Test_acc: 0.842500
Train acc: 0.928000, Test_acc: 0.848750
Train acc: 0.930208, Test_acc: 0.848750
Train acc: 0.928125, Test_acc: 0.846250
Train acc: 0.929417, Test_acc: 0.846250
Train acc: 0.930667, Test_acc: 0.846250
Train acc: 0.930583, Test_acc: 0.845000
Train acc: 0.930875, Test_acc: 0.846250
Train acc: 0.931000, Test_acc: 0.842500
Train acc: 0.931667, Test_acc: 0.841250
Train acc: 0.931583, Test_acc: 0.843750
Train acc: 0.932042, Test_acc: 0.848750
Train acc: 0.933250, Test_acc: 0.845000
Time elapsed - 13.929224014282227 seconds.

(-0.4 , +0.4) 1 hidden layer and 50 units:

In [ ]:
Train acc: 0.631542, Test_acc: 0.756250
Train acc: 0.811375, Test_acc: 0.808750
Train acc: 0.847875, Test_acc: 0.825000
Train acc: 0.865875, Test_acc: 0.837500
Train acc: 0.872958, Test_acc: 0.837500
Train acc: 0.878083, Test_acc: 0.845000
Train acc: 0.883458, Test_acc: 0.855000
Train acc: 0.886333, Test_acc: 0.856250
Train acc: 0.888792, Test_acc: 0.846250
Train acc: 0.892917, Test_acc: 0.858750
Train acc: 0.895250, Test_acc: 0.860000
Train acc: 0.896208, Test_acc: 0.858750
Train acc: 0.897875, Test_acc: 0.852500
Train acc: 0.898875, Test_acc: 0.853750
Train acc: 0.901125, Test_acc: 0.851250
Train acc: 0.901042, Test_acc: 0.856250
Train acc: 0.903250, Test_acc: 0.857500
Train acc: 0.903333, Test_acc: 0.851250
Train acc: 0.904042, Test_acc: 0.852500
Train acc: 0.905583, Test_acc: 0.851250
Train acc: 0.906208, Test_acc: 0.857500
Train acc: 0.907333, Test_acc: 0.845000
Train acc: 0.908167, Test_acc: 0.848750
Train acc: 0.908417, Test_acc: 0.856250
Train acc: 0.909542, Test_acc: 0.855000
Train acc: 0.909083, Test_acc: 0.846250
Train acc: 0.909083, Test_acc: 0.847500
Train acc: 0.908792, Test_acc: 0.852500
Train acc: 0.909667, Test_acc: 0.843750
Train acc: 0.909917, Test_acc: 0.848750
Train acc: 0.912208, Test_acc: 0.843750
Train acc: 0.911792, Test_acc: 0.845000
Train acc: 0.912500, Test_acc: 0.845000
Train acc: 0.912458, Test_acc: 0.851250
Train acc: 0.913167, Test_acc: 0.838750
Train acc: 0.913458, Test_acc: 0.842500
Train acc: 0.915167, Test_acc: 0.843750
Train acc: 0.914875, Test_acc: 0.845000
Train acc: 0.915917, Test_acc: 0.852500
Train acc: 0.916250, Test_acc: 0.853750
Train acc: 0.915958, Test_acc: 0.852500
Train acc: 0.917250, Test_acc: 0.845000
Train acc: 0.917042, Test_acc: 0.843750
Train acc: 0.917042, Test_acc: 0.843750
Train acc: 0.916875, Test_acc: 0.845000
Train acc: 0.915750, Test_acc: 0.838750
Train acc: 0.916083, Test_acc: 0.847500
Train acc: 0.916958, Test_acc: 0.846250
Train acc: 0.917750, Test_acc: 0.848750
Train acc: 0.916583, Test_acc: 0.852500
Time elapsed - 15.104513168334961 seconds.

(-0.4 , +0.4) 2 layers and 10 units for each layer: 

In [ ]:
Train acc: 0.613875, Test_acc: 0.760000
Train acc: 0.824042, Test_acc: 0.817500
Train acc: 0.863292, Test_acc: 0.847500
Train acc: 0.881583, Test_acc: 0.848750
Train acc: 0.892542, Test_acc: 0.853750
Train acc: 0.901292, Test_acc: 0.852500
Train acc: 0.907083, Test_acc: 0.857500
Train acc: 0.911875, Test_acc: 0.863750
Train acc: 0.914792, Test_acc: 0.866250
Train acc: 0.917333, Test_acc: 0.863750
Train acc: 0.919125, Test_acc: 0.861250
Train acc: 0.921458, Test_acc: 0.861250
Train acc: 0.922417, Test_acc: 0.855000
Train acc: 0.922542, Test_acc: 0.852500
Train acc: 0.924250, Test_acc: 0.852500
Train acc: 0.924625, Test_acc: 0.852500
Train acc: 0.924292, Test_acc: 0.853750
Train acc: 0.925667, Test_acc: 0.851250
Train acc: 0.926792, Test_acc: 0.850000
Train acc: 0.927500, Test_acc: 0.850000
Train acc: 0.929042, Test_acc: 0.856250
Train acc: 0.932083, Test_acc: 0.853750
Train acc: 0.934625, Test_acc: 0.853750
Train acc: 0.936250, Test_acc: 0.857500
Train acc: 0.937208, Test_acc: 0.858750
Train acc: 0.938000, Test_acc: 0.855000
Train acc: 0.938667, Test_acc: 0.858750
Train acc: 0.939250, Test_acc: 0.858750
Train acc: 0.939542, Test_acc: 0.857500
Train acc: 0.939875, Test_acc: 0.857500
Train acc: 0.939958, Test_acc: 0.857500
Train acc: 0.940167, Test_acc: 0.858750
Train acc: 0.940250, Test_acc: 0.857500
Train acc: 0.940375, Test_acc: 0.860000
Train acc: 0.940583, Test_acc: 0.860000
Train acc: 0.940708, Test_acc: 0.861250
Train acc: 0.940833, Test_acc: 0.860000
Train acc: 0.941083, Test_acc: 0.861250
Train acc: 0.941375, Test_acc: 0.861250
Train acc: 0.941417, Test_acc: 0.860000
Train acc: 0.941583, Test_acc: 0.861250
Train acc: 0.941625, Test_acc: 0.860000
Train acc: 0.941750, Test_acc: 0.860000
Train acc: 0.941792, Test_acc: 0.857500
Train acc: 0.941958, Test_acc: 0.857500
Train acc: 0.942000, Test_acc: 0.857500
Train acc: 0.937417, Test_acc: 0.860000
Train acc: 0.931167, Test_acc: 0.855000
Train acc: 0.931458, Test_acc: 0.846250
Train acc: 0.934167, Test_acc: 0.850000
Time elapsed - 14.201554298400879 seconds.

(-0.4 , +0.4) 2 hidden layers and 50 units each layer:

In [ ]:
Train acc: 0.602083, Test_acc: 0.746250
Train acc: 0.823500, Test_acc: 0.803750
Train acc: 0.865375, Test_acc: 0.826250
Train acc: 0.885000, Test_acc: 0.823750
Train acc: 0.898167, Test_acc: 0.840000
Train acc: 0.908542, Test_acc: 0.833750
Train acc: 0.916667, Test_acc: 0.840000
Train acc: 0.922375, Test_acc: 0.843750
Train acc: 0.928250, Test_acc: 0.840000
Train acc: 0.932250, Test_acc: 0.832500
Train acc: 0.935458, Test_acc: 0.835000
Train acc: 0.939083, Test_acc: 0.840000
Train acc: 0.942417, Test_acc: 0.836250
Train acc: 0.944917, Test_acc: 0.848750
Train acc: 0.946750, Test_acc: 0.843750
Train acc: 0.948542, Test_acc: 0.847500
Train acc: 0.950500, Test_acc: 0.842500
Train acc: 0.951750, Test_acc: 0.842500
Train acc: 0.952083, Test_acc: 0.840000
Train acc: 0.953125, Test_acc: 0.843750
Train acc: 0.953833, Test_acc: 0.840000
Train acc: 0.954792, Test_acc: 0.837500
Train acc: 0.955083, Test_acc: 0.841250
Train acc: 0.955792, Test_acc: 0.826250
Train acc: 0.957417, Test_acc: 0.826250
Train acc: 0.956417, Test_acc: 0.830000
Train acc: 0.958917, Test_acc: 0.826250
Train acc: 0.958458, Test_acc: 0.833750
Train acc: 0.959167, Test_acc: 0.838750
Train acc: 0.960500, Test_acc: 0.832500
Train acc: 0.962708, Test_acc: 0.836250
Train acc: 0.964167, Test_acc: 0.845000
Train acc: 0.965875, Test_acc: 0.843750
Train acc: 0.966458, Test_acc: 0.838750
Train acc: 0.967583, Test_acc: 0.841250
Train acc: 0.968542, Test_acc: 0.841250
Train acc: 0.968833, Test_acc: 0.842500
Train acc: 0.969458, Test_acc: 0.842500
Train acc: 0.969500, Test_acc: 0.823750
Train acc: 0.970208, Test_acc: 0.826250
Train acc: 0.970417, Test_acc: 0.835000
Train acc: 0.970500, Test_acc: 0.831250
Train acc: 0.971000, Test_acc: 0.822500
Train acc: 0.971292, Test_acc: 0.826250
Train acc: 0.971625, Test_acc: 0.818750
Train acc: 0.972042, Test_acc: 0.826250
Train acc: 0.972083, Test_acc: 0.830000
Train acc: 0.972125, Test_acc: 0.840000
Train acc: 0.972542, Test_acc: 0.836250
Train acc: 0.972208, Test_acc: 0.828750
Time elapsed - 16.029121160507202 seconds.

100 iteration (-0.4 , +0.4) 2 hidden layers and 50 units each layer:

In [ ]:
Train acc: 0.690125, Test_acc: 0.752500
Train acc: 0.819583, Test_acc: 0.803750
Train acc: 0.860125, Test_acc: 0.807500
Train acc: 0.880917, Test_acc: 0.813750
Train acc: 0.893125, Test_acc: 0.823750
Train acc: 0.905125, Test_acc: 0.830000
Train acc: 0.913167, Test_acc: 0.821250
Train acc: 0.919500, Test_acc: 0.823750
Train acc: 0.926125, Test_acc: 0.822500
Train acc: 0.931625, Test_acc: 0.826250
Train acc: 0.935375, Test_acc: 0.833750
Train acc: 0.938250, Test_acc: 0.842500
Train acc: 0.940750, Test_acc: 0.841250
Train acc: 0.944250, Test_acc: 0.842500
Train acc: 0.945917, Test_acc: 0.846250
Train acc: 0.949208, Test_acc: 0.848750
Train acc: 0.951583, Test_acc: 0.845000
Train acc: 0.953792, Test_acc: 0.847500
Train acc: 0.955292, Test_acc: 0.846250
Train acc: 0.956417, Test_acc: 0.838750
Train acc: 0.958208, Test_acc: 0.836250
Train acc: 0.960708, Test_acc: 0.843750
Train acc: 0.961250, Test_acc: 0.845000
Train acc: 0.962375, Test_acc: 0.843750
Train acc: 0.962542, Test_acc: 0.841250
Train acc: 0.963333, Test_acc: 0.827500
Train acc: 0.963333, Test_acc: 0.845000
Train acc: 0.964167, Test_acc: 0.831250
Train acc: 0.965250, Test_acc: 0.837500
Train acc: 0.965583, Test_acc: 0.833750
Train acc: 0.964708, Test_acc: 0.845000
Train acc: 0.965583, Test_acc: 0.836250
Train acc: 0.966917, Test_acc: 0.835000
Train acc: 0.967000, Test_acc: 0.832500
Train acc: 0.968083, Test_acc: 0.833750
Train acc: 0.969208, Test_acc: 0.837500
Train acc: 0.970708, Test_acc: 0.833750
Train acc: 0.970083, Test_acc: 0.841250
Train acc: 0.969958, Test_acc: 0.836250
Train acc: 0.969708, Test_acc: 0.830000
Train acc: 0.969625, Test_acc: 0.826250
Train acc: 0.970708, Test_acc: 0.832500
Train acc: 0.970125, Test_acc: 0.827500
Train acc: 0.971458, Test_acc: 0.833750
Train acc: 0.972917, Test_acc: 0.827500
Train acc: 0.972500, Test_acc: 0.828750
Train acc: 0.973708, Test_acc: 0.835000
Train acc: 0.974000, Test_acc: 0.831250
Train acc: 0.973750, Test_acc: 0.828750
Train acc: 0.974208, Test_acc: 0.828750
Train acc: 0.975083, Test_acc: 0.821250
Train acc: 0.975417, Test_acc: 0.831250
Train acc: 0.975917, Test_acc: 0.818750
Train acc: 0.975917, Test_acc: 0.820000
Train acc: 0.976167, Test_acc: 0.818750
Train acc: 0.976292, Test_acc: 0.828750
Train acc: 0.976625, Test_acc: 0.833750
Train acc: 0.976292, Test_acc: 0.833750
Train acc: 0.975875, Test_acc: 0.825000
Train acc: 0.975583, Test_acc: 0.821250
Train acc: 0.973125, Test_acc: 0.830000
Train acc: 0.968458, Test_acc: 0.820000
Train acc: 0.963083, Test_acc: 0.833750
Train acc: 0.961250, Test_acc: 0.831250
Train acc: 0.961333, Test_acc: 0.840000
Train acc: 0.964750, Test_acc: 0.826250
Train acc: 0.962583, Test_acc: 0.833750
Train acc: 0.965292, Test_acc: 0.831250
Train acc: 0.967167, Test_acc: 0.827500
Train acc: 0.969083, Test_acc: 0.827500
Train acc: 0.970917, Test_acc: 0.833750
Train acc: 0.971042, Test_acc: 0.828750
Train acc: 0.972667, Test_acc: 0.835000
Train acc: 0.974208, Test_acc: 0.838750
Train acc: 0.974792, Test_acc: 0.840000
Train acc: 0.973875, Test_acc: 0.830000
Train acc: 0.975250, Test_acc: 0.830000
Train acc: 0.975875, Test_acc: 0.833750
Train acc: 0.977167, Test_acc: 0.832500
Train acc: 0.977833, Test_acc: 0.828750
Train acc: 0.977750, Test_acc: 0.835000
Train acc: 0.978417, Test_acc: 0.833750
Train acc: 0.978583, Test_acc: 0.836250
Train acc: 0.978583, Test_acc: 0.836250
Train acc: 0.978917, Test_acc: 0.836250
Train acc: 0.978875, Test_acc: 0.832500
Train acc: 0.979083, Test_acc: 0.831250
Train acc: 0.979208, Test_acc: 0.832500
Train acc: 0.979292, Test_acc: 0.831250
Train acc: 0.979292, Test_acc: 0.831250
Train acc: 0.979333, Test_acc: 0.831250
Train acc: 0.979375, Test_acc: 0.831250
Train acc: 0.979375, Test_acc: 0.833750
Train acc: 0.979417, Test_acc: 0.832500
Train acc: 0.979417, Test_acc: 0.831250
Train acc: 0.979417, Test_acc: 0.827500
Train acc: 0.979500, Test_acc: 0.830000
Train acc: 0.979542, Test_acc: 0.828750
Train acc: 0.979625, Test_acc: 0.827500
Train acc: 0.979625, Test_acc: 0.827500
Time elapsed - 30.76068902015686 seconds.

### Report 2:

We can use (-0.4 , +0.4) 1 hidden layer and 10 units to compare approach 1 since they use the same database of "useful words" to classify reviews. Approcah 1 has 76% and approach 2 has 93% training accuracy and 84% testing accuracy. In approach 2, the neural network is adjusting weights every iteration, but approach 1 does not enhance base on testing data.   

### Report 3:

We put many sets of result data to make it easier to compare both accuracy and running time under different situations. We first compare data of (-0 , +0) and (-0.4 , +0.4) for 1 hidden layer and 10 units, we can see that (-0.4 , 0.4) is much faster than (-0.1 , 0.1) because when we set a low ignore-word range, we are leaving much more words to the input vector so that the matrix multiplication $w^Tx$ will take more time. However, we can see that their final accuracy is very close and both are higher than approach 1. To both (-0 , +0) and (-0.4 , +0.4), when they have 1 layer and increase their units to 50 in those layers, both of their final training accuracy get decreased but performance of testing accuracy varies. When (-0 , +0) and (-0.4 , +0.4) has two layers and 10 units each layer, we can see the both running time and accuracy are getting better, which means if there are two strategies, having a single layer with many units, and having more layers and less units, we may want to use the second strategy. And increasing units for multiple layers may not increase testing accuracy significantly but will definitely increase running time by few seconds. Also as we can see, when we increase number of epochs, complex models as two layers neural network with 50 units each layer will have almost the same accuracy compare to the result when there are only 50 epochs.

### Part 2:
#### Output:

In [ ]:
1 layer 28*28

Epoch 1/500
100000/100000 [==============================] - 14s 140us/step - loss: 3.1568 - accuracy: 0.6073
Epoch 2/500
100000/100000 [==============================] - 14s 136us/step - loss: 1.4298 - accuracy: 0.6920
Epoch 3/500
100000/100000 [==============================] - 14s 138us/step - loss: 1.4020 - accuracy: 0.7101
Epoch 4/500
100000/100000 [==============================] - 15s 150us/step - loss: 1.3966 - accuracy: 0.7182
Epoch 5/500
100000/100000 [==============================] - 16s 161us/step - loss: 1.3710 - accuracy: 0.7235
Epoch 6/500
100000/100000 [==============================] - 17s 171us/step - loss: 1.3756 - accuracy: 0.7269
Epoch 7/500
100000/100000 [==============================] - 16s 156us/step - loss: 1.3680 - accuracy: 0.7304
Epoch 8/500
100000/100000 [==============================] - 15s 145us/step - loss: 1.3694 - accuracy: 0.7364
Epoch 9/500
100000/100000 [==============================] - 13s 134us/step - loss: 1.3415 - accuracy: 0.7403
Epoch 10/5

100000/100000 [==============================] - 13s 127us/step - loss: 0.8774 - accuracy: 0.8009
Epoch 148/500
100000/100000 [==============================] - 13s 135us/step - loss: 0.8915 - accuracy: 0.7997
Epoch 149/500
100000/100000 [==============================] - 14s 140us/step - loss: 0.8979 - accuracy: 0.8004
Epoch 150/500
100000/100000 [==============================] - 13s 134us/step - loss: 0.8756 - accuracy: 0.8005
Epoch 151/500
100000/100000 [==============================] - 14s 137us/step - loss: 0.8594 - accuracy: 0.7991
Epoch 152/500
100000/100000 [==============================] - 15s 147us/step - loss: 0.8748 - accuracy: 0.8006
Epoch 153/500
100000/100000 [==============================] - 14s 139us/step - loss: 0.8779 - accuracy: 0.8014
Epoch 154/500
100000/100000 [==============================] - 12s 119us/step - loss: 0.8950 - accuracy: 0.7997
Epoch 155/500
100000/100000 [==============================] - 12s 125us/step - loss: 0.8828 - accuracy: 0.8004
Epoch 

100000/100000 [==============================] - 13s 129us/step - loss: 0.8524 - accuracy: 0.7947
Epoch 221/500
100000/100000 [==============================] - 13s 127us/step - loss: 0.8514 - accuracy: 0.7950
Epoch 222/500
100000/100000 [==============================] - 13s 128us/step - loss: 0.8520 - accuracy: 0.7954
Epoch 223/500
100000/100000 [==============================] - 12s 125us/step - loss: 0.8598 - accuracy: 0.7948
Epoch 224/500
100000/100000 [==============================] - 12s 124us/step - loss: 0.8683 - accuracy: 0.7945
Epoch 225/500
100000/100000 [==============================] - 13s 130us/step - loss: 0.8426 - accuracy: 0.7953
Epoch 226/500
100000/100000 [==============================] - 13s 127us/step - loss: 0.8587 - accuracy: 0.7957
Epoch 227/500
100000/100000 [==============================] - 13s 127us/step - loss: 0.8493 - accuracy: 0.7950
Epoch 228/500
100000/100000 [==============================] - 12s 122us/step - loss: 0.8635 - accuracy: 0.7953
Epoch 

100000/100000 [==============================] - 13s 129us/step - loss: 0.7939 - accuracy: 0.7986
Epoch 367/500
100000/100000 [==============================] - 13s 126us/step - loss: 0.7939 - accuracy: 0.7987
Epoch 368/500
100000/100000 [==============================] - 14s 141us/step - loss: 0.7685 - accuracy: 0.7991
Epoch 369/500
100000/100000 [==============================] - 12s 123us/step - loss: 0.7836 - accuracy: 0.7986
Epoch 370/500
100000/100000 [==============================] - 12s 124us/step - loss: 0.7700 - accuracy: 0.7993
Epoch 371/500
100000/100000 [==============================] - 13s 127us/step - loss: 0.7845 - accuracy: 0.7990
Epoch 372/500
100000/100000 [==============================] - 13s 127us/step - loss: 0.7881 - accuracy: 0.7984
Epoch 373/500
100000/100000 [==============================] - 13s 126us/step - loss: 0.7818 - accuracy: 0.7991
Epoch 374/500
100000/100000 [==============================] - 12s 121us/step - loss: 0.7827 - accuracy: 0.7997
Epoch 

In [ ]:
1 layer 20*20

Epoch 1/500
100000/100000 [==============================] - 10s 101us/step - loss: 8.5626 - acc: 0.4642
Epoch 2/500
100000/100000 [==============================] - 7s 69us/step - loss: 7.5545 - acc: 0.5284
Epoch 3/500
100000/100000 [==============================] - 6s 61us/step - loss: 7.2726 - acc: 0.5465
Epoch 4/500
100000/100000 [==============================] - 6s 61us/step - loss: 7.1099 - acc: 0.5568
Epoch 5/500
100000/100000 [==============================] - 6s 61us/step - loss: 6.8602 - acc: 0.5725
Epoch 6/500
100000/100000 [==============================] - 6s 62us/step - loss: 6.7289 - acc: 0.5805
Epoch 7/500
100000/100000 [==============================] - 7s 66us/step - loss: 6.5741 - acc: 0.5904
Epoch 8/500
100000/100000 [==============================] - 6s 61us/step - loss: 6.6354 - acc: 0.5865
Epoch 9/500
100000/100000 [==============================] - 6s 61us/step - loss: 6.4816 - acc: 0.5960
Epoch 10/500
100000/100000 [==============================] - 6s 61us/s

100000/100000 [==============================] - 8s 75us/step - loss: 5.1081 - acc: 0.6825
Epoch 158/500
100000/100000 [==============================] - 8s 80us/step - loss: 5.1304 - acc: 0.6811
Epoch 159/500
100000/100000 [==============================] - 8s 77us/step - loss: 5.1095 - acc: 0.6824
Epoch 160/500
100000/100000 [==============================] - 8s 78us/step - loss: 5.1035 - acc: 0.6828
Epoch 161/500
100000/100000 [==============================] - 8s 76us/step - loss: 5.0960 - acc: 0.6833
Epoch 162/500
100000/100000 [==============================] - 8s 84us/step - loss: 5.1426 - acc: 0.6804
Epoch 163/500
100000/100000 [==============================] - 8s 80us/step - loss: 5.1147 - acc: 0.6822
Epoch 164/500
100000/100000 [==============================] - 8s 79us/step - loss: 5.1313 - acc: 0.6811
Epoch 165/500
100000/100000 [==============================] - 8s 84us/step - loss: 5.1132 - acc: 0.6823
Epoch 166/500
100000/100000 [==============================] - 8s 79u

100000/100000 [==============================] - 8s 82us/step - loss: 4.9604 - acc: 0.6918
Epoch 236/500
100000/100000 [==============================] - 9s 91us/step - loss: 4.9782 - acc: 0.6906
Epoch 237/500
100000/100000 [==============================] - 9s 90us/step - loss: 4.9655 - acc: 0.6915
Epoch 238/500
100000/100000 [==============================] - 9s 87us/step - loss: 4.9778 - acc: 0.6906
Epoch 239/500
100000/100000 [==============================] - 9s 85us/step - loss: 4.9805 - acc: 0.6905
Epoch 240/500
100000/100000 [==============================] - 9s 89us/step - loss: 5.0264 - acc: 0.6877
Epoch 241/500
100000/100000 [==============================] - 9s 92us/step - loss: 5.0043 - acc: 0.6890
Epoch 242/500
100000/100000 [==============================] - 9s 87us/step - loss: 5.0073 - acc: 0.6888
Epoch 243/500
100000/100000 [==============================] - 9s 90us/step - loss: 4.9723 - acc: 0.6911
Epoch 244/500
100000/100000 [==============================] - 9s 89u

100000/100000 [==============================] - 10s 102us/step - loss: 4.8508 - acc: 0.6986
Epoch 388/500
100000/100000 [==============================] - ETA: 0s - loss: 4.8419 - acc: 0.699 - 11s 107us/step - loss: 4.8412 - acc: 0.6993
Epoch 389/500
100000/100000 [==============================] - 11s 113us/step - loss: 4.8374 - acc: 0.6995
Epoch 390/500
100000/100000 [==============================] - 11s 106us/step - loss: 4.8331 - acc: 0.6998
Epoch 391/500
100000/100000 [==============================] - 11s 110us/step - loss: 4.8513 - acc: 0.6986
Epoch 392/500
100000/100000 [==============================] - 11s 105us/step - loss: 4.8385 - acc: 0.6995
Epoch 393/500
100000/100000 [==============================] - 11s 110us/step - loss: 4.8242 - acc: 0.7003
Epoch 394/500
100000/100000 [==============================] - 12s 121us/step - loss: 4.8255 - acc: 0.7003
Epoch 395/500
100000/100000 [==============================] - 9s 93us/step - loss: 4.8093 - acc: 0.7013
Epoch 396/500
1

In [ ]:
1 layer 15*15

Epoch 1/500
100000/100000 [==============================] - 6s 63us/step - loss: 8.6663 - acc: 0.4584
Epoch 2/500
100000/100000 [==============================] - 6s 55us/step - loss: 8.0276 - acc: 0.5001
Epoch 3/500
100000/100000 [==============================] - 6s 61us/step - loss: 7.9560 - acc: 0.5049
Epoch 4/500
100000/100000 [==============================] - 6s 56us/step - loss: 7.9227 - acc: 0.5072
Epoch 5/500
100000/100000 [==============================] - 6s 57us/step - loss: 7.8794 - acc: 0.5101
Epoch 6/500
100000/100000 [==============================] - 6s 58us/step - loss: 7.8501 - acc: 0.5121
Epoch 7/500
100000/100000 [==============================] - 6s 56us/step - loss: 7.8291 - acc: 0.5135
Epoch 8/500
100000/100000 [==============================] - 6s 60us/step - loss: 7.8252 - acc: 0.5138
Epoch 9/500
100000/100000 [==============================] - 6s 56us/step - loss: 7.8265 - acc: 0.5137
Epoch 10/500
100000/100000 [==============================] - 6s 56us/ste

100000/100000 [==============================] - 5s 53us/step - loss: 7.3465 - acc: 0.5439
Epoch 158/500
100000/100000 [==============================] - 5s 52us/step - loss: 7.3394 - acc: 0.5444
Epoch 159/500
100000/100000 [==============================] - 5s 55us/step - loss: 7.3965 - acc: 0.5409
Epoch 160/500
100000/100000 [==============================] - 5s 52us/step - loss: 7.3439 - acc: 0.5442
Epoch 161/500
100000/100000 [==============================] - 5s 53us/step - loss: 7.2992 - acc: 0.5469
Epoch 162/500
100000/100000 [==============================] - 6s 58us/step - loss: 7.4040 - acc: 0.5404
Epoch 163/500
100000/100000 [==============================] - 5s 52us/step - loss: 7.3098 - acc: 0.5463
Epoch 164/500
100000/100000 [==============================] - 5s 53us/step - loss: 7.3299 - acc: 0.5451
Epoch 165/500
100000/100000 [==============================] - 5s 54us/step - loss: 7.3240 - acc: 0.5455
Epoch 166/500
100000/100000 [==============================] - 6s 56u

100000/100000 [==============================] - 5s 55us/step - loss: 7.1854 - acc: 0.5541
Epoch 312/500
100000/100000 [==============================] - 6s 57us/step - loss: 7.1997 - acc: 0.5532
Epoch 313/500
100000/100000 [==============================] - 6s 57us/step - loss: 7.2530 - acc: 0.5498
Epoch 314/500
100000/100000 [==============================] - 5s 54us/step - loss: 7.2166 - acc: 0.5521
Epoch 315/500
100000/100000 [==============================] - 5s 54us/step - loss: 7.2078 - acc: 0.5527
Epoch 316/500
100000/100000 [==============================] - 6s 55us/step - loss: 7.1890 - acc: 0.5538
Epoch 317/500
100000/100000 [==============================] - 6s 55us/step - loss: 7.1930 - acc: 0.5535
Epoch 318/500
100000/100000 [==============================] - 6s 55us/step - loss: 7.2106 - acc: 0.5525
Epoch 319/500
100000/100000 [==============================] - 6s 64us/step - loss: 7.1801 - acc: 0.5544
Epoch 320/500
100000/100000 [==============================] - 6s 56u

100000/100000 [==============================] - 6s 61us/step - loss: 7.1523 - acc: 0.5562
Epoch 467/500
100000/100000 [==============================] - 7s 65us/step - loss: 7.1745 - acc: 0.5547
Epoch 468/500
100000/100000 [==============================] - 7s 66us/step - loss: 7.1643 - acc: 0.5554
Epoch 469/500
100000/100000 [==============================] - 5s 53us/step - loss: 7.1420 - acc: 0.5568
Epoch 470/500
100000/100000 [==============================] - 6s 59us/step - loss: 7.1418 - acc: 0.5568
Epoch 471/500
100000/100000 [==============================] - 6s 62us/step - loss: 7.1496 - acc: 0.5564
Epoch 472/500
100000/100000 [==============================] - 6s 56us/step - loss: 7.1727 - acc: 0.5549
Epoch 473/500
100000/100000 [==============================] - 6s 61us/step - loss: 7.1361 - acc: 0.5571
Epoch 474/500
100000/100000 [==============================] - 6s 58us/step - loss: 7.1474 - acc: 0.5565
Epoch 475/500
100000/100000 [==============================] - 7s 73u

In [ ]:
1 layer 10*10

Epoch 1/500
100000/100000 [==============================] - 7s 69us/step - loss: 7.8174 - acc: 0.4982
Epoch 2/500
100000/100000 [==============================] - 6s 56us/step - loss: 6.7222 - acc: 0.5721
Epoch 3/500
100000/100000 [==============================] - 5s 55us/step - loss: 6.4209 - acc: 0.5918
Epoch 4/500
100000/100000 [==============================] - 5s 55us/step - loss: 6.1233 - acc: 0.6107
Epoch 5/500
100000/100000 [==============================] - 5s 53us/step - loss: 6.0326 - acc: 0.6177
Epoch 6/500
100000/100000 [==============================] - 5s 54us/step - loss: 5.9207 - acc: 0.6249
Epoch 7/500
100000/100000 [==============================] - 6s 62us/step - loss: 5.7342 - acc: 0.6363
Epoch 8/500
100000/100000 [==============================] - 5s 52us/step - loss: 5.6721 - acc: 0.6406
Epoch 9/500
100000/100000 [==============================] - 6s 58us/step - loss: 5.6639 - acc: 0.6417
Epoch 10/500
100000/100000 [==============================] - 6s 62us/ste

100000/100000 [==============================] - 5s 47us/step - loss: 5.0987 - acc: 0.6824
Epoch 144/500
100000/100000 [==============================] - ETA: 0s - loss: 5.0401 - acc: 0.686 - 5s 46us/step - loss: 5.0391 - acc: 0.6861
Epoch 145/500
100000/100000 [==============================] - 5s 46us/step - loss: 5.0647 - acc: 0.6847
Epoch 146/500
100000/100000 [==============================] - 5s 46us/step - loss: 5.1054 - acc: 0.6822
Epoch 147/500
100000/100000 [==============================] - 5s 46us/step - loss: 5.0878 - acc: 0.6833
Epoch 148/500
100000/100000 [==============================] - 5s 53us/step - loss: 5.0512 - acc: 0.6854
Epoch 149/500
100000/100000 [==============================] - 5s 47us/step - loss: 5.0366 - acc: 0.6864
Epoch 150/500
100000/100000 [==============================] - 5s 47us/step - loss: 5.0744 - acc: 0.6842
Epoch 151/500
100000/100000 [==============================] - 6s 59us/step - loss: 5.0766 - acc: 0.6841
Epoch 152/500
100000/100000 [==

100000/100000 [==============================] - 6s 58us/step - loss: 5.2324 - acc: 0.6747
Epoch 299/500
100000/100000 [==============================] - 5s 54us/step - loss: 5.2557 - acc: 0.6733
Epoch 300/500
100000/100000 [==============================] - 5s 47us/step - loss: 5.2185 - acc: 0.6755
Epoch 301/500
100000/100000 [==============================] - 6s 58us/step - loss: 5.2296 - acc: 0.6749
Epoch 302/500
100000/100000 [==============================] - 5s 47us/step - loss: 5.3273 - acc: 0.6689
Epoch 303/500
100000/100000 [==============================] - ETA: 0s - loss: 5.2322 - acc: 0.674 - 5s 46us/step - loss: 5.2323 - acc: 0.6747
Epoch 304/500
100000/100000 [==============================] - 5s 47us/step - loss: 5.2435 - acc: 0.6738
Epoch 305/500
100000/100000 [==============================] - 5s 52us/step - loss: 5.1891 - acc: 0.6774
Epoch 306/500
100000/100000 [==============================] - 5s 48us/step - loss: 5.1811 - acc: 0.6780
Epoch 307/500
100000/100000 [==

100000/100000 [==============================] - 5s 47us/step - loss: 5.2222 - acc: 0.6756
Epoch 454/500
100000/100000 [==============================] - 5s 50us/step - loss: 5.2980 - acc: 0.6708
Epoch 455/500
100000/100000 [==============================] - 5s 47us/step - loss: 5.2234 - acc: 0.6754
Epoch 456/500
100000/100000 [==============================] - 5s 48us/step - loss: 5.2789 - acc: 0.6721
Epoch 457/500
100000/100000 [==============================] - 5s 46us/step - loss: 5.2439 - acc: 0.6742
Epoch 458/500
100000/100000 [==============================] - 5s 52us/step - loss: 5.2374 - acc: 0.6746
Epoch 459/500
100000/100000 [==============================] - 5s 50us/step - loss: 5.2515 - acc: 0.6737
Epoch 460/500
100000/100000 [==============================] - 5s 51us/step - loss: 5.2686 - acc: 0.6726
Epoch 461/500
100000/100000 [==============================] - 5s 47us/step - loss: 5.2185 - acc: 0.6759
Epoch 462/500
100000/100000 [==============================] - 5s 46u

In [ ]:
1 layer 5*5

Epoch 1/500
100000/100000 [==============================] - 6s 62us/step - loss: 9.7652 - acc: 0.3539
Epoch 2/500
100000/100000 [==============================] - 6s 65us/step - loss: 8.4159 - acc: 0.3962
Epoch 3/500
100000/100000 [==============================] - 4s 42us/step - loss: 6.3367 - acc: 0.3911
Epoch 4/500
100000/100000 [==============================] - 4s 42us/step - loss: 5.8044 - acc: 0.4104
Epoch 5/500
100000/100000 [==============================] - 5s 49us/step - loss: 5.7899 - acc: 0.4173
Epoch 6/500
100000/100000 [==============================] - 4s 45us/step - loss: 4.1998 - acc: 0.4497
Epoch 7/500
100000/100000 [==============================] - 4s 42us/step - loss: 1.7380 - acc: 0.5223
Epoch 8/500
100000/100000 [==============================] - 6s 64us/step - loss: 1.6273 - acc: 0.5335: 0s - loss: 1.6297 - ac
Epoch 9/500
100000/100000 [==============================] - 5s 47us/step - loss: 1.6178 - acc: 0.5333
Epoch 10/500
100000/100000 [=====================

100000/100000 [==============================] - 5s 53us/step - loss: 7.4225 - acc: 0.5376
Epoch 158/500
100000/100000 [==============================] - 5s 46us/step - loss: 7.4954 - acc: 0.5330
Epoch 159/500
100000/100000 [==============================] - 5s 47us/step - loss: 7.4116 - acc: 0.5381
Epoch 160/500
100000/100000 [==============================] - 6s 60us/step - loss: 7.3279 - acc: 0.5435
Epoch 161/500
100000/100000 [==============================] - 5s 46us/step - loss: 7.3798 - acc: 0.5401
Epoch 162/500
100000/100000 [==============================] - 5s 46us/step - loss: 7.3874 - acc: 0.5398
Epoch 163/500
100000/100000 [==============================] - 5s 49us/step - loss: 7.4141 - acc: 0.5381
Epoch 164/500
100000/100000 [==============================] - 5s 53us/step - loss: 7.4584 - acc: 0.5353
Epoch 165/500
100000/100000 [==============================] - 5s 47us/step - loss: 7.4276 - acc: 0.5371
Epoch 166/500
100000/100000 [==============================] - 5s 47u

100000/100000 [==============================] - 5s 53us/step - loss: 7.5070 - acc: 0.5331
Epoch 313/500
100000/100000 [==============================] - 5s 47us/step - loss: 7.4994 - acc: 0.5336
Epoch 314/500
100000/100000 [==============================] - 5s 47us/step - loss: 7.4911 - acc: 0.5342
Epoch 315/500
100000/100000 [==============================] - 6s 60us/step - loss: 7.4738 - acc: 0.5353
Epoch 316/500
100000/100000 [==============================] - 5s 46us/step - loss: 7.4873 - acc: 0.5344
Epoch 317/500
100000/100000 [==============================] - 5s 46us/step - loss: 7.5043 - acc: 0.5335
Epoch 318/500
100000/100000 [==============================] - 5s 45us/step - loss: 7.5165 - acc: 0.5326
Epoch 319/500
100000/100000 [==============================] - 5s 53us/step - loss: 7.5837 - acc: 0.5286
Epoch 320/500
100000/100000 [==============================] - 5s 49us/step - loss: 7.5609 - acc: 0.5301
Epoch 321/500
100000/100000 [==============================] - 5s 46u

100000/100000 [==============================] - 5s 48us/step - loss: 7.5696 - acc: 0.5297: 1s - l
Epoch 468/500
100000/100000 [==============================] - 5s 55us/step - loss: 7.5696 - acc: 0.5296
Epoch 469/500
100000/100000 [==============================] - 5s 54us/step - loss: 7.5525 - acc: 0.5306
Epoch 470/500
100000/100000 [==============================] - 5s 47us/step - loss: 7.6092 - acc: 0.5272
Epoch 471/500
100000/100000 [==============================] - 6s 61us/step - loss: 7.5700 - acc: 0.5297
Epoch 472/500
100000/100000 [==============================] - 5s 48us/step - loss: 7.5826 - acc: 0.5290
Epoch 473/500
100000/100000 [==============================] - 5s 47us/step - loss: 7.5515 - acc: 0.5309
Epoch 474/500
100000/100000 [==============================] - 5s 53us/step - loss: 7.5117 - acc: 0.5333
Epoch 475/500
100000/100000 [==============================] - 5s 47us/step - loss: 7.5388 - acc: 0.5317
Epoch 476/500
100000/100000 [==============================] 

In [ ]:
2 layers 28*28

Epoch 1/500
100000/100000 [==============================] - 10s 98us/step - loss: 14.3548 - acc: 0.1092
Epoch 2/500
100000/100000 [==============================] - 7s 74us/step - loss: 13.0303 - acc: 0.1911
Epoch 3/500
100000/100000 [==============================] - 8s 80us/step - loss: 12.0771 - acc: 0.2505
Epoch 4/500
100000/100000 [==============================] - 8s 78us/step - loss: 11.9529 - acc: 0.2583
Epoch 5/500
100000/100000 [==============================] - 7s 74us/step - loss: 11.9060 - acc: 0.2612
Epoch 6/500
100000/100000 [==============================] - 8s 75us/step - loss: 11.8775 - acc: 0.2630
Epoch 7/500
100000/100000 [==============================] - 8s 83us/step - loss: 11.8626 - acc: 0.2640
Epoch 8/500
100000/100000 [==============================] - 7s 75us/step - loss: 11.8898 - acc: 0.2623
Epoch 9/500
100000/100000 [==============================] - 8s 75us/step - loss: 11.8813 - acc: 0.2628
Epoch 10/500
100000/100000 [==============================] - 8

100000/100000 [==============================] - 8s 82us/step - loss: 10.3414 - acc: 0.3584
Epoch 143/500
100000/100000 [==============================] - 9s 88us/step - loss: 10.2749 - acc: 0.3625
Epoch 144/500
100000/100000 [==============================] - 9s 88us/step - loss: 10.2844 - acc: 0.3619
Epoch 145/500
100000/100000 [==============================] - 8s 79us/step - loss: 10.3830 - acc: 0.3558
Epoch 146/500
100000/100000 [==============================] - 8s 80us/step - loss: 10.3135 - acc: 0.3601
Epoch 147/500
100000/100000 [==============================] - 8s 79us/step - loss: 10.2677 - acc: 0.3629
Epoch 148/500
100000/100000 [==============================] - 9s 86us/step - loss: 10.2566 - acc: 0.3636
Epoch 149/500
100000/100000 [==============================] - 8s 85us/step - loss: 10.2651 - acc: 0.3631
Epoch 150/500
100000/100000 [==============================] - 8s 79us/step - loss: 10.3031 - acc: 0.3608
Epoch 151/500
100000/100000 [==============================]

100000/100000 [==============================] - 8s 79us/step - loss: 10.2973 - acc: 0.3611
Epoch 297/500
100000/100000 [==============================] - 8s 78us/step - loss: 10.3408 - acc: 0.3584
Epoch 298/500
100000/100000 [==============================] - 9s 86us/step - loss: 10.2401 - acc: 0.3647
Epoch 299/500
100000/100000 [==============================] - 8s 81us/step - loss: 10.2708 - acc: 0.3628
Epoch 300/500
100000/100000 [==============================] - 8s 80us/step - loss: 10.2751 - acc: 0.3625
Epoch 301/500
100000/100000 [==============================] - 8s 80us/step - loss: 10.2566 - acc: 0.3637
Epoch 302/500
100000/100000 [==============================] - 9s 85us/step - loss: 10.2368 - acc: 0.3648
Epoch 303/500
100000/100000 [==============================] - 8s 80us/step - loss: 10.2430 - acc: 0.3645
Epoch 304/500
100000/100000 [==============================] - 10s 96us/step - loss: 10.2356 - acc: 0.3649
Epoch 305/500
100000/100000 [==============================

100000/100000 [==============================] - 8s 82us/step - loss: 10.2139 - acc: 0.3663
Epoch 451/500
100000/100000 [==============================] - 8s 77us/step - loss: 10.2139 - acc: 0.3663
Epoch 452/500
100000/100000 [==============================] - 8s 76us/step - loss: 10.2139 - acc: 0.3663
Epoch 453/500
100000/100000 [==============================] - 8s 77us/step - loss: 10.2139 - acc: 0.3663
Epoch 454/500
100000/100000 [==============================] - 8s 79us/step - loss: 10.2139 - acc: 0.3663
Epoch 455/500
100000/100000 [==============================] - 8s 77us/step - loss: 10.2139 - acc: 0.3663
Epoch 456/500
100000/100000 [==============================] - 8s 77us/step - loss: 10.2139 - acc: 0.3663
Epoch 457/500
100000/100000 [==============================] - 9s 86us/step - loss: 10.2139 - acc: 0.3663
Epoch 458/500
100000/100000 [==============================] - 8s 75us/step - loss: 10.2139 - acc: 0.3663
Epoch 459/500
100000/100000 [==============================]

In [ ]:
2 layers 20*20

Epoch 1/500
100000/100000 [==============================] - 9s 85us/step - loss: 11.7015 - acc: 0.2725
Epoch 2/500
100000/100000 [==============================] - 9s 95us/step - loss: 10.7854 - acc: 0.3301
Epoch 3/500
100000/100000 [==============================] - 8s 83us/step - loss: 10.7137 - acc: 0.3347
Epoch 4/500
100000/100000 [==============================] - 11s 106us/step - loss: 10.6369 - acc: 0.3396
Epoch 5/500
100000/100000 [==============================] - 8s 85us/step - loss: 10.6162 - acc: 0.3410
Epoch 6/500
100000/100000 [==============================] - 8s 78us/step - loss: 10.5913 - acc: 0.3426
Epoch 7/500
100000/100000 [==============================] - 8s 84us/step - loss: 10.5822 - acc: 0.3431
Epoch 8/500
100000/100000 [==============================] - 8s 81us/step - loss: 10.6269 - acc: 0.3405
Epoch 9/500
100000/100000 [==============================] - 8s 81us/step - loss: 10.6091 - acc: 0.3416
Epoch 10/500
100000/100000 [==============================] - 

Epoch 79/500
100000/100000 [==============================] - 8s 83us/step - loss: 10.3598 - acc: 0.3572
Epoch 80/500
100000/100000 [==============================] - 8s 84us/step - loss: 10.3979 - acc: 0.3548
Epoch 81/500
100000/100000 [==============================] - 9s 87us/step - loss: 10.3640 - acc: 0.3569
Epoch 82/500
100000/100000 [==============================] - 8s 85us/step - loss: 10.4186 - acc: 0.3536
Epoch 83/500
100000/100000 [==============================] - 8s 83us/step - loss: 10.3732 - acc: 0.3564
Epoch 84/500
100000/100000 [==============================] - 9s 88us/step - loss: 10.3539 - acc: 0.3576
Epoch 85/500
100000/100000 [==============================] - 8s 83us/step - loss: 10.3943 - acc: 0.3551
Epoch 86/500
100000/100000 [==============================] - 8s 83us/step - loss: 10.3582 - acc: 0.3573
Epoch 87/500
100000/100000 [==============================] - 8s 83us/step - loss: 10.3748 - acc: 0.3563
Epoch 88/500
100000/100000 [===========================

100000/100000 [==============================] - 11s 110us/step - loss: 10.3129 - acc: 0.3601
Epoch 233/500
100000/100000 [==============================] - 10s 104us/step - loss: 10.3176 - acc: 0.3599
Epoch 234/500
100000/100000 [==============================] - 10s 104us/step - loss: 10.3015 - acc: 0.3609
Epoch 235/500
100000/100000 [==============================] - 11s 110us/step - loss: 10.3388 - acc: 0.3585
Epoch 236/500
100000/100000 [==============================] - 11s 106us/step - loss: 10.4178 - acc: 0.3536
Epoch 237/500
100000/100000 [==============================] - 11s 106us/step - loss: 10.3617 - acc: 0.3571
Epoch 238/500
100000/100000 [==============================] - 11s 109us/step - loss: 10.3498 - acc: 0.3579
Epoch 239/500
100000/100000 [==============================] - 10s 104us/step - loss: 10.3206 - acc: 0.3597
Epoch 240/500
100000/100000 [==============================] - 11s 105us/step - loss: 10.2865 - acc: 0.3618
Epoch 241/500
100000/100000 [=============

Epoch 308/500
100000/100000 [==============================] - 15s 151us/step - loss: 10.3545 - acc: 0.3575
Epoch 309/500
100000/100000 [==============================] - 13s 126us/step - loss: 10.3039 - acc: 0.3607
Epoch 310/500
100000/100000 [==============================] - 14s 141us/step - loss: 10.3214 - acc: 0.3596
Epoch 311/500
100000/100000 [==============================] - 15s 149us/step - loss: 10.2947 - acc: 0.3613
Epoch 312/500
100000/100000 [==============================] - 12s 119us/step - loss: 10.3028 - acc: 0.3608
Epoch 313/500
100000/100000 [==============================] - 13s 134us/step - loss: 10.2697 - acc: 0.3628
Epoch 314/500
100000/100000 [==============================] - 12s 124us/step - loss: 10.3033 - acc: 0.3607
Epoch 315/500
100000/100000 [==============================] - 11s 113us/step - loss: 10.3342 - acc: 0.3588
Epoch 316/500
100000/100000 [==============================] - 13s 132us/step - loss: 10.3187 - acc: 0.3598
Epoch 317/500
100000/100000 

Epoch 459/500
100000/100000 [==============================] - 13s 131us/step - loss: 10.2505 - acc: 0.3640
Epoch 460/500
100000/100000 [==============================] - 13s 125us/step - loss: 10.2571 - acc: 0.3636
Epoch 461/500
100000/100000 [==============================] - 13s 131us/step - loss: 10.2421 - acc: 0.3645
Epoch 462/500
100000/100000 [==============================] - 13s 127us/step - loss: 10.2661 - acc: 0.3631
Epoch 463/500
100000/100000 [==============================] - 13s 131us/step - loss: 10.2666 - acc: 0.3630
Epoch 464/500
100000/100000 [==============================] - 12s 125us/step - loss: 10.2387 - acc: 0.3647
Epoch 465/500
100000/100000 [==============================] - 13s 130us/step - loss: 10.2604 - acc: 0.3634
Epoch 466/500
100000/100000 [==============================] - 13s 130us/step - loss: 10.2908 - acc: 0.3615
Epoch 467/500
100000/100000 [==============================] - 13s 127us/step - loss: 10.2515 - acc: 0.3639
Epoch 468/500
100000/100000 

In [ ]:
2 layers and resize to 15*15

Epoch 1/500
100000/100000 [==============================] - 8s 76us/step - loss: 13.0429 - acc: 0.1904
Epoch 2/500
100000/100000 [==============================] - 6s 61us/step - loss: 12.9902 - acc: 0.1940
Epoch 3/500
100000/100000 [==============================] - 7s 73us/step - loss: 12.9815 - acc: 0.1945
Epoch 4/500
100000/100000 [==============================] - 6s 59us/step - loss: 11.9202 - acc: 0.2597
Epoch 5/500
100000/100000 [==============================] - 6s 63us/step - loss: 10.8951 - acc: 0.3234
Epoch 6/500
100000/100000 [==============================] - 6s 56us/step - loss: 10.9526 - acc: 0.3196
Epoch 7/500
100000/100000 [==============================] - 6s 58us/step - loss: 10.3375 - acc: 0.3575
Epoch 8/500
100000/100000 [==============================] - 6s 56us/step - loss: 9.9616 - acc: 0.3811
Epoch 9/500
100000/100000 [==============================] - 6s 57us/step - loss: 9.7834 - acc: 0.3923
Epoch 10/500
100000/100000 [==============================] - 6s 6

100000/100000 [==============================] - 9s 89us/step - loss: 9.1022 - acc: 0.4352
Epoch 158/500
100000/100000 [==============================] - 6s 64us/step - loss: 9.1172 - acc: 0.4343
Epoch 159/500
100000/100000 [==============================] - 6s 61us/step - loss: 9.1942 - acc: 0.4295
Epoch 160/500
100000/100000 [==============================] - 6s 64us/step - loss: 9.1153 - acc: 0.4344
Epoch 161/500
100000/100000 [==============================] - 7s 66us/step - loss: 9.1166 - acc: 0.4343
Epoch 162/500
100000/100000 [==============================] - 7s 74us/step - loss: 9.0990 - acc: 0.4354
Epoch 163/500
100000/100000 [==============================] - 8s 75us/step - loss: 9.1430 - acc: 0.4327
Epoch 164/500
100000/100000 [==============================] - 7s 69us/step - loss: 9.0573 - acc: 0.4380
Epoch 165/500
100000/100000 [==============================] - 7s 69us/step - loss: 9.0668 - acc: 0.4374
Epoch 166/500
100000/100000 [==============================] - 7s 75u

100000/100000 [==============================] - 7s 70us/step - loss: 8.9746 - acc: 0.4432
Epoch 314/500
100000/100000 [==============================] - 8s 81us/step - loss: 8.9607 - acc: 0.4440
Epoch 315/500
100000/100000 [==============================] - 7s 70us/step - loss: 8.9593 - acc: 0.4441
Epoch 316/500
100000/100000 [==============================] - 7s 71us/step - loss: 8.9432 - acc: 0.4451
Epoch 317/500
100000/100000 [==============================] - 7s 70us/step - loss: 8.9999 - acc: 0.4415
Epoch 318/500
100000/100000 [==============================] - 8s 77us/step - loss: 8.9789 - acc: 0.4429
Epoch 319/500
100000/100000 [==============================] - 7s 70us/step - loss: 8.9804 - acc: 0.4428
Epoch 320/500
100000/100000 [==============================] - 7s 71us/step - loss: 8.9597 - acc: 0.4441
Epoch 321/500
100000/100000 [==============================] - 7s 70us/step - loss: 8.9646 - acc: 0.4437
Epoch 322/500
100000/100000 [==============================] - 7s 72u

100000/100000 [==============================] - 7s 69us/step - loss: 8.9506 - acc: 0.4447
Epoch 470/500
100000/100000 [==============================] - 7s 69us/step - loss: 8.9458 - acc: 0.4449
Epoch 471/500
100000/100000 [==============================] - 7s 73us/step - loss: 8.9247 - acc: 0.4462
Epoch 472/500
100000/100000 [==============================] - 7s 69us/step - loss: 8.9096 - acc: 0.4472
Epoch 473/500
100000/100000 [==============================] - 7s 68us/step - loss: 8.9492 - acc: 0.4448
Epoch 474/500
100000/100000 [==============================] - 7s 67us/step - loss: 8.9411 - acc: 0.4453
Epoch 475/500
100000/100000 [==============================] - 9s 85us/step - loss: 8.9437 - acc: 0.4451
Epoch 476/500
100000/100000 [==============================] - 7s 71us/step - loss: 8.9308 - acc: 0.4459
Epoch 477/500
100000/100000 [==============================] - 7s 70us/step - loss: 8.9260 - acc: 0.4462
Epoch 478/500
100000/100000 [==============================] - 8s 81u

In [ ]:
2 layer 10*10

Epoch 1/500
100000/100000 [==============================] - 7s 65us/step - loss: 12.3898 - acc: 0.2283
Epoch 2/500
100000/100000 [==============================] - 7s 70us/step - loss: 10.7222 - acc: 0.3310
Epoch 3/500
100000/100000 [==============================] - 6s 55us/step - loss: 10.2079 - acc: 0.3641
Epoch 4/500
100000/100000 [==============================] - 5s 54us/step - loss: 9.9766 - acc: 0.3787
Epoch 5/500
100000/100000 [==============================] - 6s 59us/step - loss: 9.9365 - acc: 0.3818
Epoch 6/500
100000/100000 [==============================] - 7s 73us/step - loss: 9.8526 - acc: 0.3873
Epoch 7/500
100000/100000 [==============================] - 7s 71us/step - loss: 9.8224 - acc: 0.3890
Epoch 8/500
100000/100000 [==============================] - 6s 60us/step - loss: 9.8198 - acc: 0.3896
Epoch 9/500
100000/100000 [==============================] - 7s 72us/step - loss: 9.7936 - acc: 0.3912
Epoch 10/500
100000/100000 [==============================] - 6s 56us/

100000/100000 [==============================] - 6s 58us/step - loss: 9.1742 - acc: 0.4305
Epoch 158/500
100000/100000 [==============================] - 6s 58us/step - loss: 9.2652 - acc: 0.4249
Epoch 159/500
100000/100000 [==============================] - 6s 58us/step - loss: 9.1949 - acc: 0.4292
Epoch 160/500
100000/100000 [==============================] - 6s 58us/step - loss: 9.2023 - acc: 0.4288
Epoch 161/500
100000/100000 [==============================] - 6s 64us/step - loss: 9.1596 - acc: 0.4315
Epoch 162/500
100000/100000 [==============================] - 6s 59us/step - loss: 9.1980 - acc: 0.4290
Epoch 163/500
100000/100000 [==============================] - 6s 61us/step - loss: 9.1979 - acc: 0.4291
Epoch 164/500
100000/100000 [==============================] - 6s 58us/step - loss: 9.1792 - acc: 0.4302
Epoch 165/500
100000/100000 [==============================] - 6s 59us/step - loss: 9.1750 - acc: 0.4305
Epoch 166/500
100000/100000 [==============================] - 6s 63u

100000/100000 [==============================] - 5s 54us/step - loss: 9.0565 - acc: 0.4380
Epoch 314/500
100000/100000 [==============================] - 5s 53us/step - loss: 9.0600 - acc: 0.4377
Epoch 315/500
100000/100000 [==============================] - 6s 61us/step - loss: 9.0575 - acc: 0.4379
Epoch 316/500
100000/100000 [==============================] - 5s 52us/step - loss: 9.1018 - acc: 0.4351
Epoch 317/500
100000/100000 [==============================] - 5s 52us/step - loss: 9.1012 - acc: 0.4351
Epoch 318/500
100000/100000 [==============================] - 5s 53us/step - loss: 9.1172 - acc: 0.4341
Epoch 319/500
100000/100000 [==============================] - 5s 52us/step - loss: 9.0818 - acc: 0.4363
Epoch 320/500
100000/100000 [==============================] - 6s 59us/step - loss: 9.0821 - acc: 0.4363
Epoch 321/500
100000/100000 [==============================] - 5s 53us/step - loss: 9.0749 - acc: 0.4368
Epoch 322/500
100000/100000 [==============================] - 5s 54u

100000/100000 [==============================] - 5s 53us/step - loss: 9.0380 - acc: 0.4391
Epoch 470/500
100000/100000 [==============================] - 5s 50us/step - loss: 9.0247 - acc: 0.4400
Epoch 471/500
100000/100000 [==============================] - 5s 52us/step - loss: 9.0122 - acc: 0.4407
Epoch 472/500
100000/100000 [==============================] - 5s 53us/step - loss: 9.0279 - acc: 0.4397
Epoch 473/500
100000/100000 [==============================] - 7s 70us/step - loss: 9.0179 - acc: 0.4404
Epoch 474/500
100000/100000 [==============================] - 5s 52us/step - loss: 9.0268 - acc: 0.4398
Epoch 475/500
100000/100000 [==============================] - 5s 51us/step - loss: 9.0228 - acc: 0.4400
Epoch 476/500
100000/100000 [==============================] - 6s 56us/step - loss: 9.0227 - acc: 0.4401
Epoch 477/500
100000/100000 [==============================] - 5s 53us/step - loss: 9.0320 - acc: 0.4395
Epoch 478/500
100000/100000 [==============================] - 5s 53u

In [ ]:
2 layers 5*5

Epoch 1/500
100000/100000 [==============================] - 9s 85us/step - loss: 10.7426 - acc: 0.3107
Epoch 2/500
100000/100000 [==============================] - 7s 65us/step - loss: 2.2927 - acc: 0.4705
Epoch 3/500
100000/100000 [==============================] - 7s 72us/step - loss: 1.4378 - acc: 0.5326
Epoch 4/500
100000/100000 [==============================] - 6s 60us/step - loss: 1.3931 - acc: 0.5462
Epoch 5/500
100000/100000 [==============================] - 6s 60us/step - loss: 1.3785 - acc: 0.5571
Epoch 6/500
100000/100000 [==============================] - 6s 65us/step - loss: 1.3819 - acc: 0.5604
Epoch 7/500
100000/100000 [==============================] - 7s 66us/step - loss: 1.3689 - acc: 0.5625
Epoch 8/500
100000/100000 [==============================] - 6s 57us/step - loss: 1.3604 - acc: 0.5657
Epoch 9/500
100000/100000 [==============================] - 7s 74us/step - loss: 1.3615 - acc: 0.5661
Epoch 10/500
100000/100000 [==============================] - 7s 71us/st

100000/100000 [==============================] - 7s 66us/step - loss: 1.7958 - acc: 0.5577
Epoch 144/500
100000/100000 [==============================] - 11s 109us/step - loss: 1.8032 - acc: 0.5583
Epoch 145/500
100000/100000 [==============================] - 5s 52us/step - loss: 1.8256 - acc: 0.5585
Epoch 146/500
100000/100000 [==============================] - 5s 50us/step - loss: 1.7953 - acc: 0.5595
Epoch 147/500
100000/100000 [==============================] - 7s 65us/step - loss: 1.8287 - acc: 0.5585
Epoch 148/500
100000/100000 [==============================] - 5s 54us/step - loss: 1.8510 - acc: 0.5575
Epoch 149/500
100000/100000 [==============================] - 6s 56us/step - loss: 1.9341 - acc: 0.5545
Epoch 150/500
100000/100000 [==============================] - 5s 50us/step - loss: 1.8897 - acc: 0.5560
Epoch 151/500
100000/100000 [==============================] - 7s 67us/step - loss: 1.9184 - acc: 0.5535
Epoch 152/500
100000/100000 [==============================] - 5s 5

100000/100000 [==============================] - 6s 56us/step - loss: 2.1346 - acc: 0.5474
Epoch 300/500
100000/100000 [==============================] - 5s 54us/step - loss: 2.1552 - acc: 0.5476
Epoch 301/500
100000/100000 [==============================] - 6s 58us/step - loss: 2.1023 - acc: 0.5489
Epoch 302/500
100000/100000 [==============================] - 5s 53us/step - loss: 2.1261 - acc: 0.5471
Epoch 303/500
100000/100000 [==============================] - 5s 53us/step - loss: 2.1133 - acc: 0.5485
Epoch 304/500
100000/100000 [==============================] - 5s 54us/step - loss: 2.1295 - acc: 0.5482
Epoch 305/500
100000/100000 [==============================] - 5s 54us/step - loss: 2.1403 - acc: 0.5493
Epoch 306/500
100000/100000 [==============================] - 6s 59us/step - loss: 2.1227 - acc: 0.5476
Epoch 307/500
100000/100000 [==============================] - 5s 54us/step - loss: 2.1097 - acc: 0.5488
Epoch 308/500
100000/100000 [==============================] - 5s 55u

100000/100000 [==============================] - 5s 55us/step - loss: 2.5245 - acc: 0.5194
Epoch 456/500
100000/100000 [==============================] - 5s 53us/step - loss: 2.5535 - acc: 0.5168
Epoch 457/500
100000/100000 [==============================] - 5s 53us/step - loss: 2.5307 - acc: 0.5198
Epoch 458/500
100000/100000 [==============================] - 6s 59us/step - loss: 2.5676 - acc: 0.5189
Epoch 459/500
100000/100000 [==============================] - 5s 54us/step - loss: 2.5682 - acc: 0.5198
Epoch 460/500
100000/100000 [==============================] - 6s 56us/step - loss: 2.4950 - acc: 0.5203
Epoch 461/500
100000/100000 [==============================] - 6s 55us/step - loss: 2.4936 - acc: 0.5186
Epoch 462/500
100000/100000 [==============================] - 6s 55us/step - loss: 2.4677 - acc: 0.5185
Epoch 463/500
100000/100000 [==============================] - 6s 57us/step - loss: 2.4794 - acc: 0.5189
Epoch 464/500
100000/100000 [==============================] - 6s 63u

In [ ]:
3 layers 28*28

Epoch 1/500
100000/100000 [==============================] - 20s 196us/step - loss: 1.5963 - accuracy: 0.6727
Epoch 2/500
100000/100000 [==============================] - 20s 198us/step - loss: 0.9608 - accuracy: 0.7483
Epoch 3/500
100000/100000 [==============================] - 19s 187us/step - loss: 1.0629 - accuracy: 0.7287
Epoch 4/500
100000/100000 [==============================] - 18s 185us/step - loss: 1.1491 - accuracy: 0.7142
Epoch 5/500
100000/100000 [==============================] - 18s 180us/step - loss: 1.4167 - accuracy: 0.6681
Epoch 6/500
100000/100000 [==============================] - 19s 188us/step - loss: 1.5980 - accuracy: 0.6227
Epoch 7/500
100000/100000 [==============================] - 18s 185us/step - loss: 1.6892 - accuracy: 0.5904
Epoch 8/500
100000/100000 [==============================] - 19s 188us/step - loss: 1.9254 - accuracy: 0.5788
Epoch 9/500
100000/100000 [==============================] - 19s 187us/step - loss: 2.1655 - accuracy: 0.5367
Epoch 10/5

100000/100000 [==============================] - 18s 183us/step - loss: 3.5294 - accuracy: 0.1883
Epoch 148/500
100000/100000 [==============================] - 18s 177us/step - loss: 2.7776 - accuracy: 0.1850
Epoch 149/500
100000/100000 [==============================] - 18s 177us/step - loss: 2.4708 - accuracy: 0.1752
Epoch 150/500
100000/100000 [==============================] - 17s 175us/step - loss: 2.5510 - accuracy: 0.1734
Epoch 151/500
100000/100000 [==============================] - 17s 175us/step - loss: 2.6310 - accuracy: 0.1867
Epoch 152/500
100000/100000 [==============================] - 17s 174us/step - loss: 2.5243 - accuracy: 0.1764
Epoch 153/500
100000/100000 [==============================] - 17s 174us/step - loss: 2.3036 - accuracy: 0.1643
Epoch 154/500
100000/100000 [==============================] - 17s 173us/step - loss: 2.2285 - accuracy: 0.1714
Epoch 155/500
100000/100000 [==============================] - 18s 176us/step - loss: 2.4896 - accuracy: 0.1797
Epoch 

100000/100000 [==============================] - 19s 186us/step - loss: 2.2513 - accuracy: 0.1878
Epoch 293/500
100000/100000 [==============================] - 19s 186us/step - loss: 2.2609 - accuracy: 0.1773
Epoch 294/500
100000/100000 [==============================] - 19s 187us/step - loss: 2.1887 - accuracy: 0.1835
Epoch 295/500
100000/100000 [==============================] - 19s 186us/step - loss: 2.1648 - accuracy: 0.1834
Epoch 296/500
100000/100000 [==============================] - 19s 186us/step - loss: 2.1994 - accuracy: 0.1802
Epoch 297/500
100000/100000 [==============================] - 18s 185us/step - loss: 2.2007 - accuracy: 0.1804
Epoch 298/500
100000/100000 [==============================] - 19s 187us/step - loss: 2.2092 - accuracy: 0.1830
Epoch 299/500
100000/100000 [==============================] - 23s 225us/step - loss: 2.2139 - accuracy: 0.1854
Epoch 300/500
100000/100000 [==============================] - 20s 195us/step - loss: 2.2123 - accuracy: 0.1868
Epoch 

100000/100000 [==============================] - 24s 240us/step - loss: 2.1390 - accuracy: 0.2115
Epoch 437/500
100000/100000 [==============================] - 18s 179us/step - loss: 2.1689 - accuracy: 0.2098
Epoch 438/500
100000/100000 [==============================] - 19s 188us/step - loss: 2.2313 - accuracy: 0.2096
Epoch 439/500
100000/100000 [==============================] - 19s 189us/step - loss: 2.1602 - accuracy: 0.2045
Epoch 440/500
100000/100000 [==============================] - 20s 203us/step - loss: 2.0825 - accuracy: 0.2084
Epoch 441/500
100000/100000 [==============================] - 18s 176us/step - loss: 2.0991 - accuracy: 0.2104
Epoch 442/500
100000/100000 [==============================] - 22s 221us/step - loss: 2.1924 - accuracy: 0.2089
Epoch 443/500
100000/100000 [==============================] - 23s 230us/step - loss: 2.0444 - accuracy: 0.2083
Epoch 444/500
100000/100000 [==============================] - 25s 249us/step - loss: 2.2147 - accuracy: 0.2068
Epoch 

In [ ]:
3 layers 28*28 with 64 nodes per layer

Epoch 1/500
100000/100000 [==============================] - 6s 57us/step - loss: 10.6624 - acc: 0.3357
Epoch 2/500
100000/100000 [==============================] - 6s 63us/step - loss: 8.7719 - acc: 0.4539
Epoch 3/500
100000/100000 [==============================] - 5s 54us/step - loss: 8.3415 - acc: 0.4812
Epoch 4/500
100000/100000 [==============================] - 5s 52us/step - loss: 8.1722 - acc: 0.4918
Epoch 5/500
100000/100000 [==============================] - 5s 54us/step - loss: 8.0962 - acc: 0.4969
Epoch 6/500
100000/100000 [==============================] - 5s 52us/step - loss: 8.0612 - acc: 0.4993
Epoch 7/500
100000/100000 [==============================] - 5s 53us/step - loss: 8.0758 - acc: 0.4983
Epoch 8/500
100000/100000 [==============================] - 6s 60us/step - loss: 8.0687 - acc: 0.4988
Epoch 9/500
100000/100000 [==============================] - 5s 52us/step - loss: 7.9682 - acc: 0.5051
Epoch 10/500
100000/100000 [==============================] - 6s 58us/st

100000/100000 [==============================] - 7s 70us/step - loss: 7.5637 - acc: 0.5307
Epoch 158/500
100000/100000 [==============================] - 6s 56us/step - loss: 7.5528 - acc: 0.5313
Epoch 159/500
100000/100000 [==============================] - 5s 54us/step - loss: 7.5540 - acc: 0.5313
Epoch 160/500
100000/100000 [==============================] - 6s 57us/step - loss: 7.5657 - acc: 0.5305
Epoch 161/500
100000/100000 [==============================] - 6s 61us/step - loss: 7.5793 - acc: 0.5297
Epoch 162/500
100000/100000 [==============================] - 6s 55us/step - loss: 7.5754 - acc: 0.5300
Epoch 163/500
100000/100000 [==============================] - 6s 57us/step - loss: 7.5642 - acc: 0.5306
Epoch 164/500
100000/100000 [==============================] - 5s 55us/step - loss: 7.5548 - acc: 0.5312
Epoch 165/500
100000/100000 [==============================] - 5s 55us/step - loss: 7.6519 - acc: 0.5252
Epoch 166/500
100000/100000 [==============================] - 6s 59u

100000/100000 [==============================] - 6s 63us/step - loss: 7.4656 - acc: 0.5368
Epoch 236/500
100000/100000 [==============================] - 6s 55us/step - loss: 7.4768 - acc: 0.5361
Epoch 237/500
100000/100000 [==============================] - 5s 55us/step - loss: 7.5205 - acc: 0.5334
Epoch 238/500
100000/100000 [==============================] - 5s 54us/step - loss: 7.5458 - acc: 0.5318
Epoch 239/500
100000/100000 [==============================] - 5s 55us/step - loss: 7.5739 - acc: 0.5300
Epoch 240/500
100000/100000 [==============================] - 6s 56us/step - loss: 7.5216 - acc: 0.5333
Epoch 241/500
100000/100000 [==============================] - 6s 63us/step - loss: 7.4750 - acc: 0.5362
Epoch 242/500
100000/100000 [==============================] - 5s 54us/step - loss: 7.6061 - acc: 0.5280
Epoch 243/500
100000/100000 [==============================] - 9s 93us/step - loss: 7.5382 - acc: 0.5323
Epoch 244/500
100000/100000 [==============================] - 6s 55u

100000/100000 [==============================] - 5s 50us/step - loss: 7.4515 - acc: 0.5377
Epoch 391/500
100000/100000 [==============================] - 5s 51us/step - loss: 7.4661 - acc: 0.5367
Epoch 392/500
100000/100000 [==============================] - 5s 51us/step - loss: 7.4375 - acc: 0.5385
Epoch 393/500
100000/100000 [==============================] - 6s 58us/step - loss: 7.4933 - acc: 0.5351
Epoch 394/500
100000/100000 [==============================] - 5s 50us/step - loss: 7.4058 - acc: 0.5405
Epoch 395/500
100000/100000 [==============================] - 5s 51us/step - loss: 7.4452 - acc: 0.5381
Epoch 396/500
100000/100000 [==============================] - 5s 52us/step - loss: 7.4113 - acc: 0.5402
Epoch 397/500
100000/100000 [==============================] - 5s 50us/step - loss: 7.4386 - acc: 0.5385
Epoch 398/500
100000/100000 [==============================] - 5s 51us/step - loss: 7.4724 - acc: 0.5364
Epoch 399/500
100000/100000 [==============================] - 6s 56u

In [ ]:
5 layers 28*28

Epoch 1/500
100000/100000 [==============================] - 16s 157us/step - loss: 14.5060 - acc: 0.1000
Epoch 2/500
100000/100000 [==============================] - 15s 148us/step - loss: 14.5063 - acc: 0.1000
Epoch 3/500
100000/100000 [==============================] - 15s 151us/step - loss: 14.5063 - acc: 0.1000
Epoch 4/500
100000/100000 [==============================] - 14s 142us/step - loss: 14.5063 - acc: 0.1000
Epoch 5/500
100000/100000 [==============================] - 15s 146us/step - loss: 14.5063 - acc: 0.1000
Epoch 6/500
100000/100000 [==============================] - 15s 149us/step - loss: 14.5063 - acc: 0.1000
Epoch 7/500
100000/100000 [==============================] - 17s 167us/step - loss: 14.5063 - acc: 0.1000
Epoch 8/500
100000/100000 [==============================] - 16s 165us/step - loss: 14.5063 - acc: 0.1000
Epoch 9/500
100000/100000 [==============================] - 15s 150us/step - loss: 14.5063 - acc: 0.1000
Epoch 10/500
100000/100000 [==================

100000/100000 [==============================] - 16s 156us/step - loss: 14.5063 - acc: 0.1000
Epoch 153/500
100000/100000 [==============================] - 14s 138us/step - loss: 14.5063 - acc: 0.1000
Epoch 154/500
100000/100000 [==============================] - 15s 151us/step - loss: 14.5063 - acc: 0.1000
Epoch 155/500
100000/100000 [==============================] - 15s 153us/step - loss: 14.5063 - acc: 0.1000
Epoch 156/500
100000/100000 [==============================] - 14s 141us/step - loss: 14.5063 - acc: 0.1000
Epoch 157/500
100000/100000 [==============================] - 14s 139us/step - loss: 14.5063 - acc: 0.1000
Epoch 158/500
100000/100000 [==============================] - 14s 138us/step - loss: 14.5063 - acc: 0.1000
Epoch 159/500
100000/100000 [==============================] - 14s 139us/step - loss: 14.5063 - acc: 0.1000
Epoch 160/500
100000/100000 [==============================] - 14s 138us/step - loss: 14.5063 - acc: 0.1000
Epoch 161/500
100000/100000 [=============

100000/100000 [==============================] - 17s 170us/step - loss: 14.5063 - acc: 0.1000
Epoch 303/500
100000/100000 [==============================] - 17s 166us/step - loss: 14.5063 - acc: 0.1000
Epoch 304/500
100000/100000 [==============================] - 16s 161us/step - loss: 14.5063 - acc: 0.1000
Epoch 305/500
100000/100000 [==============================] - 19s 192us/step - loss: 14.5063 - acc: 0.1000
Epoch 306/500
100000/100000 [==============================] - 16s 160us/step - loss: 14.5063 - acc: 0.1000
Epoch 307/500
100000/100000 [==============================] - 16s 162us/step - loss: 14.5063 - acc: 0.1000
Epoch 308/500
100000/100000 [==============================] - 16s 164us/step - loss: 14.5063 - acc: 0.1000
Epoch 309/500
100000/100000 [==============================] - 15s 151us/step - loss: 14.5063 - acc: 0.1000
Epoch 310/500
100000/100000 [==============================] - 15s 152us/step - loss: 14.5063 - acc: 0.1000
Epoch 311/500
100000/100000 [=============

Epoch 378/500
100000/100000 [==============================] - 16s 155us/step - loss: 14.5063 - acc: 0.1000
Epoch 379/500
100000/100000 [==============================] - 17s 171us/step - loss: 14.5063 - acc: 0.1000
Epoch 380/500
100000/100000 [==============================] - 16s 159us/step - loss: 14.5063 - acc: 0.1000
Epoch 381/500
100000/100000 [==============================] - 18s 182us/step - loss: 14.5063 - acc: 0.1000
Epoch 382/500
100000/100000 [==============================] - 18s 185us/step - loss: 14.5063 - acc: 0.1000
Epoch 383/500
100000/100000 [==============================] - 16s 163us/step - loss: 14.5063 - acc: 0.1000
Epoch 384/500
100000/100000 [==============================] - 16s 158us/step - loss: 14.5063 - acc: 0.1000
Epoch 385/500
100000/100000 [==============================] - 16s 162us/step - loss: 14.5063 - acc: 0.1000
Epoch 386/500
100000/100000 [==============================] - 17s 167us/step - loss: 14.5063 - acc: 0.1000
Epoch 387/500
100000/100000 

### Report 4:
Accuracy and run time for all data sets we tested: 
   
1 layer 28*28:   
Training accuracy 0.80454   
Testing accuracy 0.67964   
Time elapsed - 6834.0641407 seconds.   
   
1 layer 20*20:   
Training accuracy 0.70248   
Testing accuracy 0.67664   
Time elapsed - 4504.813200712204 seconds.   
   
1 layer 15*15:   
Training accuracy 0.55451   
Testing accuracy 0.54664   
Time elapsed - 2922.5650837421417 seconds.    
   
1 layer 10*10:   
Training accuracy 0.6669   
Testing accuracy 0.6516   
Time elapsed - 2608.5697910785675 seconds.   
   
1 layer 5*5:   
Training accuracy 0.52956   
Testing accuracy 0.51836   
Time elapsed - 2611.2955887317657 seconds.   
   
2 layers 28*28:   
Training accuracy 0.36631   
Testing accuracy 0.36616   
Time elapsed - 4130.538147211075 seconds.   

2 layers 20*20:   
Training accuracy 0.36244   
Testing accuracy 0.35856   
Time elapsed - 5627.439848899841 seconds.   
   
2 layers 15*15:   
Training accuracy 0.4463   
Testing accuracy 0.43788   
Time elapsed - 3577.058732032776 seconds.   
   
2 layers 10*10:   
Training accuracy 0.43755   
Testing accuracy 0.42592   
Time elapsed - 2861.405417442322 seconds.   
   
2 layers 5*5:   
Training accuracy 0.49769   
Testing accuracy 0.481   
Time elapsed - 2907.9516365528107 seconds.   
   
3 layers 28*28:
Training accuracy 0.21544   
Testing accuracy 0.20928   
Time elapsed - 9590.078378 seconds.   
   
3 layers 28*28 with 64 nodes each layer:      
Training accuracy 0.53584   
Testing accuracy 0.5316   
Time elapsed - 2851.7370274066925 seconds.  
   
5 layers 28*28:   
Training accuracy 0.1   
Testing accuracy 0.1   
Time elapsed - 9752.742514371872 seconds.   

For this part our result for different situations are from different group members, so the running time can have some differences and it depend on a computer's performance and status while running the code. But we can still roughly conclude behaviors of different situations. When the resolution is decreased(28\*28->5\*5), the running time will be quicker but accuracy will drop when there is 1 layer. When we have two layers, the accuracies have monotonous behavior no longer but running time still decreases when resolution decreases.  
When the layer is increased from 3 to 5, the accuracy goes down significantly and running time increases. Adding layers sometimes overfit our model and make it affected too much by new data.  
Also we add a trial that use 64 nodes instead of 256 for 3 layer model. We see that simplified model may have much better accuracy in practice.  
Increasing layers in general may or may not increase accuracy and it depends a lot on the dataset itself but running time will increase. Therefore, when designing our model, we should balance the complexity of our model, which includes unit number for each layer and layer number. Sometimes complex model will have much worse accuracy and running time. 